test/requests/semantic_highlighting_expectations_test.rb

```diff
@@ -21,16 +21,75 @@ def run_expectations(source)
     listener = RubyLsp::Requests::SemanticHighlighting.new(dispatcher, range: processed_range)
 
     dispatcher.dispatch(document.tree)
-    RubyLsp::Requests::Support::SemanticTokenEncoder.new.encode(listener.perform)
+    listener.perform
   end
 
   def assert_expectations(source, expected)
     actual = run_expectations(source).data
     assert_equal(json_expectations(expected).to_json, decode_tokens(actual).to_json)
   end
 
+  def test_semantic_highlighting_addon
+    source = <<~RUBY
+      class Post
+        custom_method :foo
+        before_create :set_defaults
+      end
+    RUBY
+
+    test_addon(:create_semantic_highlighting_addon, source: source) do |executor|
+      response = executor.execute({
+        method: "textDocument/semanticTokens/full",
+        params: { textDocument: { uri: "file:///fake.rb" } },
+      })
+
+      assert_nil(response.error, response.error&.full_message)
+      decoded_response = decode_tokens(response.response.data)
+      assert_equal(
+        { delta_line: 0, delta_start_char: 6, length: 4, token_type: 2, token_modifiers: 1 },
+        decoded_response[0],
+      )
+      assert_equal(
+        { delta_line: 0, delta_start_char: 0, length: 4, token_type: 0, token_modifiers: 0 },
+        decoded_response[1],
+      )
+      assert_equal(
+        { delta_line: 1, delta_start_char: 2, length: 13, token_type: 13, token_modifiers: 0 },
+        decoded_response[2],
+      )
+      # This is the token modified by the addon
+      assert_equal(
+        { delta_line: 1, delta_start_char: 2, length: 13, token_type: 13, token_modifiers: 1 },
+        decoded_response[3],
+      )
+    end
+  end
+
   private
 
+  def create_semantic_highlighting_addon
+    Class.new(RubyLsp::Addon) do
+      def create_semantic_highlighting_listener(response_builder, dispatcher)
+        klass = Class.new do
+          include RubyLsp::Requests::Support::Common
+
+          def initialize(response_builder, dispatcher)
+            @response_builder = response_builder
+            dispatcher.register(self, :on_call_node_enter)
+          end
+
+          def on_call_node_enter(node)
+            if node.message == "before_create"
+              @response_builder.last.modifier << 0
+            end
```

Comment:
Let's add a test for an addon that needs to override the token type rather than changing the modifiers and check what happens.

We do this for the Sorbet annotations and it'll be necessary for other addons. The base Ruby LSP pushes the `method` token type for method calls, but we want the annotations to have the `type` token type.

It's important we stick to the types and modifiers from the [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokenTypes) because that's what themes use to define the colours they will use. If we deviate from that, then existing themes will highlight code incorrectly.

The main question here is: if the Ruby LSP and an addon push different token types for the same exact location, does VS Code prefer the first or the second token?

If it prefers the second one, then maybe we don't need to do anything and just allow the last addon to have the final say. If it picks the first one, then we'll need to override the type.

Reply:
> Is there ever a case where an addon would like to modify a token that isn't the current one (last one we pushed)?

I can't think of any at the moment.

> And how would the addon author know they need to modify an existing token rather than adding a separate one? Is there an API we can provide that will guide them to do the right thing?

I was thinking about not adding `add_token` to the builder and use `instance_variable_get` in our own listener to insert tokens. Then the only exposed APIs are `replace_type` and `replace_modifier`. WDYT?

Reply:
I'm not sure that would be enough. An addon could want to push a token that doesn't already exist.

What do you think about exposing an API and handling this complexity from our side? Something like this

```ruby
# This is pseudo code, but you get the idea
class SemanticHighlightingResponseBuilder
  def add_token(location, ...)
    if last_token.location == location
      # modify last token with type and modifiers
    else
      # push new token
    end
  end
end
```

Reply:
> An addon could want to push a token that doesn't already exist.

Do you have a use case for it? If not, maybe we can wait for feedback before introducing new APIs that ourselves would not need?

> What do you think about exposing an API and handling this complexity from our side?

It's worth exploring. But again, without a real use case I can not confidently say it'd be sufficient or necessary. For the ideas I have for `ruby-lsp-rails` and `ruby-lsp-rspec`, being able to replace type or modifier is already enough. Maybe when ERB support is shipped we can have new tokens? But if that's the case, let's design the API when that happens.

Also, I think we can have this discussion in a follow up issue/PR as the builder structure is needed to unblock other requests' refactors.